src/prefect/runner/runner.py

```diff
@@ -1121,3 +1137,58 @@ def my_other_flow(name):
         console.print(Panel(Group(help_message_top, table, help_message_bottom)))
 
     await runner.start()
+
+
+async def _run_hooks(
+    hooks: List[Callable[[Flow, "FlowRun", State], None]], flow_run, flow, state
+):
+    logger = flow_run_logger(flow_run, flow)
+    for hook in hooks:
+        try:
+            logger.info(
+                f"Running hook {hook.__name__!r} in response to entering state"
+                f" {state.name!r}"
+            )
+            if is_async_fn(hook):
+                await hook(flow=flow, flow_run=flow_run, state=state)
+            else:
+                await from_async.call_in_new_thread(
+                    create_call(hook, flow=flow, flow_run=flow_run, state=state)
+                )
+        except Exception:
+            logger.error(
+                f"An error was encountered while running hook {hook.__name__!r}",
+                exc_info=True,
+            )
+        else:
+            logger.info(f"Hook {hook.__name__!r} finished running successfully")
+
+
+async def _run_on_cancellation_hooks(
+    flow_run: "FlowRun",
+    state: State,
+    client: PrefectClient,
+) -> None:
+    """
+    Run the hooks for a flow.
+    """
+    if PREFECT_ENABLE_ON_CANCELLATION_HOOKS and state.is_cancelling():
+        flow = await load_flow_from_flow_run(flow_run, client=client)
+        hooks = flow.on_cancellation or []
```

Comment:
wow this looks way better than i even expected! 

note: this does prevent dynamically defined hooks per-run but I think that's perfectly acceptable; probably worth documenting as a note somewhere

Reply:
Could we have dynamic hooks defined per run for deployed flows before? I can think of how you'd be able to do that now.

Reply:
Yea I think so, because you can have a hook like:

```python
from prefect.runtime.flow_run import flow_run_name

if flow_run_name.startswith("bob"):
    def my_hook(...)
else:
    def my_hook(...)
```

which I don't think will work in any mode where the hook is called outside the process.  I think that's totally acceptable though - all dynamicism of the hook just needs to be incorporated into the function definition (which is what I think we need to call out in docs somewhere).